### PR TITLE
Spike: globally require common libraries in one place

### PR DESCRIPTION
**Pull request by harlantwood**
**This pull request is not ready to merge**, but is here to for discussion and feedback.

We require many of the same libraries in many different places.  This is a spike on gathering (most of) those into a single common location, and using the nodeJS/webpack-friendly `global` object to make them globally accessible.

Note that this is not possible in some cases, eg where a class extends another class, the superclass must be imported locally, not globally.  It does work for 80-90% of our require's though.

In this spike, the app _seems_ to work find in both dev server and when building and running the production version locally:

```
gulp build
node server.js
```

Additional testing/poking around encouraged.

[Tests pass](https://circleci.com/gh/PeggApp/pegg/130) as well.

@simple10 @auggernaut @jacksenechal @justinriggio I'm requesting feedback.  Any downsides/dangers you can see to this approach?  

If it seems good, I'll continue on, doing this at least for the files with huge blocks of requirements, to simplify things.

### CODE REVIEW

File: src/app/stores/PlayStore.coffee [hunk 1]
```diff
@@ -8,11 +8,6 @@ EventEmitter = require 'famous/src/core/EventEmitter'
 
 # Pegg
 AppDispatcher = require 'dispatchers/AppDispatcher'
```
Comment (auggernaut) [inline src/app/stores/PlayStore.coffee]: Why not remove AppDispatcher here? I see that it's in globals as well.

**Comment by auggernaut** [comment]
From Joe:

in general, i'm not a fan of adding to the global namespace. it can cause some weird side effects and also makes dependency injection unpredictable for testing, or at least it can.

another approach is to use the technique we're using for tests where there's one "common" module that's imported that exports a lot of the common modules so that the following is possible ...

{PlayStore, PlayView, Promise} = require 'lib/common'

it's really a matter of taste. but it's an anti-pattern to add to the global scope

**Comment by harlantwood** [comment]
Thanks @simple10.  In general I'm not a fan of global variables of any kind.  This seems like a good case for an exception to me, as it seems both very useful and also harmless:

1. We never expect to reuse (for example) the variable `PlayStore` anywhere in our app with any meaning other than `require 'stores/PlayStore'`.  
1. Dependency injection seems unaffected AFAICT -- if I have a local `PlayStore = fakePlayStore()` it overrides the global version.

File: src/app/stores/PlayStore.coffee (hunk 1 above)
Comment (harlantwood) [inline src/app/stores/PlayStore.coffee]: Trial and error -- I tried removing all, then put back a few which crashed ;)

**Comment by harlantwood** [comment]
Almost as convenient as the global solution, but without using globals, we could:

Have a one liner at the top of all/many files, an expanded version of this:

    {PlayStore, PlayView, Promise} = require 'lib/common'

Which included all/most common libraries.  This one liner could be exactly the same in all files, and could be search/replaced when adding a new lib to it.

This would have another advantage over the global solution: it would work in all cases, even the exceptions where globals failed (mentioned above).

**Comment by jacksenechal** [comment]
Having one line at the top of all files with ALL of the libraries in it sounds like the worst of both worlds. Not a fan of that solution. However, having a file-specific one-liner with just the ones you need could be pretty elegant.

@simple10 I'm not convinced that adding requires to the global scope is such an antipattern here... I can definitely see the problems for other types of globals, but for requires it seems clearly defined and not likely to cause problems.

**Comment by simple10** [comment]
JavaScript antipatterns is mostly a matter of taste. But in general, adding anything to the global scope makes code less portable and predictable. For instance, what happens when imbedding Pegg in Cordova or Crosswalk? Probably all good. But what if a Cordova video plugin happens to register an API called PlayView?

Here's a random StackOverflow thread on JS antipatterns: http://stackoverflow.com/questions/377999/what-anti-patterns-exist-for-javascript

Basically, it comes down to tradeoffs. Are all the requires such a pain as to warrant polluting the global space? It's up to you guys. Personal taste issue. But if you go down that route and need to revert, it's important to have close to 100% test coverage or you won't know if you forgot to restore a missing require.

My recommendation is to go the factory-factory pattern route if you really need to get rid of requires. (Personally, I'd leave the requires as they have their value.) ...

```coffeescript
playview = Factory.create 'PlayView'

# instead of
require 'PlayView'
playview = new PlayView
```

Then implement Factory.create to do the require and new object creation.

```coffeescript
class Factory
  @create: (moduleName) ->
    cls = require moduleName
    new (Function.prototype.bind.apply cls, arguments)

# Usage with arguments
playview = Factory.create 'PlayView', a, b, c
```

References:
- http://stackoverflow.com/questions/1606797/use-of-apply-with-new-operator-is-this-possible
- [Function.prototype.bind compatibility](http://kangax.github.io/compat-table/es5/)
  - alternative implementations on stackoverflow; but ecmascript 5 is probably more widely supported than famo.us

**Comment by simple10** [comment]
I got caught up geeking out on the factory-factory approach and completely forgot to mention a more obvious (and probably better) solution.

Just use modules. That's what they're there for.  ;-)

```coffeescript
# Views.coffee
# This is the Views module that references all other views
module.exports =
  PlayView: require 'views/PlayView'
  CardView: require 'views/CardView'
```

```coffeescript
# Some random file
Views = require 'Views'

play = new Views.PlayView
card = new Views.CardView
```

This approach does not pollute global object and significantly cuts down on the number of requires per file.

If you want to be really aggressive, you could put all the requires in one module....

```coffeescript
# Modules.coffee (or similar)

module.exports.Views =
  PlayView: require 'views/PlayView'
  CardView: require 'views/CardView'

module.exports.Stores =
  PlayStore: require 'stores/PlayStore'

module.exports.Famous =
  ScrollView: require 'famous/views/ScrollView'
```

Then just require the one file.

```coffeescript
# Some random file

Modules = require 'Modules'

play = Modules.Views.PlayView
```

**Comment by jacksenechal** [comment]
Ah, thanks Joe that makes a lot of sense! Modules seems like a good
solution here if we move forward with this spike.

**Comment by auggernaut** [comment]
Thanks everyone for your input, and Harlan for bringing this up and offer a solution!

+1 on the module pattern Joe describes. I went dropped a new spike with a variation on that approach:
https://github.com/PeggApp/pegg/pull/66

I like it. Clean, easy to read. Cuts back seriously on lines of code. If ya'll agree I'll go ahead and do the tedium of refactoring.

**Comment by simple10** [comment]
@auggernaut looks good. Much cleaner and actually easier to read.

**Comment by justinriggio** [comment]
I'm in the @simple10 camp here. I like the module pattern approach. +1

**Comment by harlantwood** [comment]
Thanks everyone for the lively discussion.  Closing this PR in favor of its spawn #66.

**Pull request closed without merging**